[Grid] Use a unitless spacing API (#14099)
mui-org/material-ui: docs/src/pages/layout/grid/grid.md

## Spacing

The responsive grid focuses on consistent spacing widths, rather than column width.

> Material design margins and columns follow an **8px** square baseline grid.

Reviewer: Nitpick: Should be Material Design

> The spacing property is an integer in the [0, 8] interval.

Reviewer 2: ```suggestion
The spacing property is an integer between 0 and 8.
```

Reviewer 3: However your notation is ambiguous about inclusive vs. exclusive intervals. If we can reword that to be explicit then I would agree that we should avoid mathematical notation.

Reviewer 2: > However your notation is ambiguous about inclusive vs. exclusive intervals.

Fair point, although I imagine most people would read that as inclusive. Let's be explicit though:

"The spacing property is an integer between 0 and 8 inclusive."

or

"The spacing property is an integer in the range 0 to 8."

?

Reviewer 4: I'm going with `The spacing property is an integer between 0 and 10 inclusive.`.

By default, the spacing between two grid items follows a linear function: `output(spacing) = spacing * 8px`, e.g. `spacing={2}` creates a 16px wide gap.

> This output transformation function can be customized using the theme, exactly like [the system spacing](/system/spacing/#transformation).

Reviewer 2: Is it "exactly like" the system spacing (but a different key in the theme), or actually the system spacing?

Reviewer 4: I'm going with 
`This output transformation function can be customized [using the theme](/customization/themes/#spacing).`

```jsx
  <body>
    <div style={{ padding: 20 }}>

>       <Grid container spacing={4}>

Reviewer 2: ```suggestion
      <Grid container spacing={5}>
```
